### DISCUSSION TITLE
Is it possible to create (or emulate) shear transformations?

### DISCUSSION THREAD

**Discussion by LeoCodes21** [bevyengine/bevy#14622] [Q&A] [answered]
I am trying to render 3D objects from an old game's world in Bevy. The vast majority of them have SRT (scale/rotation/translation) transformation matrices, but some of them also contain [shearing](https://en.wikipedia.org/wiki/Shear_mapping). Bevy's `Transform` component does not have a field for shear, and it doesn't seem like modifying the `GlobalTransform` (which wraps an affine matrix) has any effect.

Rendering sheared objects without the shear causes them to look out of place, so I'm wondering if there's any "accepted" way of handling this (I certainly haven't found one); even if there isn't a consensus, any suggestions are welcome.

**Comment by bugsweeper** [comment, marked as answer]
Yes, `Transform` doesn't have shear, it's rare feature, and looks like it some sort of combination of scale and rotation, you can try to calculate affine matrix by rotating till your objects diagonal would be located on one of axes, than scale (i think increasing) by that axis and rotate back. Matrix containing all that operations would be shearing matrix. Or if I'm wrong you could try to make your own shearing vertex shader with same idea in mind.

**Reply by LeoCodes21** [reply]
Thanks! I might have to explore the shader route.

**Reply by r0levrai** [reply]
I think that would do an anisotropic (= along one direction, = non-uniform) scaling, which transform each basis vector in their own direction ; not a shear, which can transform each basis vector in all other directions.
A vertex shader could do the trick, but note that it would not be a whole replacement of `Transform` (which are composed on cpu using a hierarchy).

**Reply by r0levrai** [reply]
I was wrong! I decided to test it and inverse_rotation * non-uniform-scale * rotation does indead produce some form of shear. Linear algebra is weird xD 
![image](https://github.com/user-attachments/assets/b8ab2d48-3de7-400f-887c-691cb95eaceb)
